src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs

```diff
@@ -671,11 +736,31 @@ private sealed class CoherentState
         {
             private readonly ConcurrentDictionary<string, CacheEntry> _stringEntries = new ConcurrentDictionary<string, CacheEntry>(StringKeyComparer.Instance);
             private readonly ConcurrentDictionary<object, CacheEntry> _nonStringEntries = new ConcurrentDictionary<object, CacheEntry>();
+
+#if NET9_0_OR_GREATER
+            private readonly bool _useStringAltLookup;
+            private readonly ConcurrentDictionary<string, CacheEntry>.AlternateLookup<ReadOnlySpan<char>> _stringAltLookup;
+
+            public CoherentState()
+            {
+                _useStringAltLookup = _stringEntries.TryGetAlternateLookup<ReadOnlySpan<char>>(out _stringAltLookup);
+                // we *expect* this to be available in all scenarios where this is used, but add a dev guard, and a fallback
+                Debug.Assert(_useStringAltLookup, "Expectation failure: alt-lookup feature is not available");
+            }
+#endif
+
             internal long _cacheSize;
 
             internal bool TryGetValue(object key, [NotNullWhen(true)] out CacheEntry? entry)
                 => key is string s ? _stringEntries.TryGetValue(s, out entry) : _nonStringEntries.TryGetValue(key, out entry);
 
+#if NET9_0_OR_GREATER
+            internal bool TryGetValue(ReadOnlySpan<char> key, [NotNullWhen(true)] out CacheEntry? entry)
+                => _useStringAltLookup ? _stringAltLookup.TryGetValue(key, out entry)
+                    : _stringEntries.TryGetValue(new string(key), out entry); // <== we do not expect this path to be hit; chaos fallback only
```

Comment:
Why add code for a case that can never happen? Shouldn't this just always use _stringAltLookup?

Reply:
It was meant defensively, but: if this API is reliable (in this context), fine: I'll remove the extra field and the switches

Reply:
4f601494c54bc080224f8b2bb540a004ae0da751